internal/statemachine/helper.go

```diff
@@ -4,11 +4,248 @@ import (
+			}
+			if err := helper.RunScript(hookScriptPath); err != nil {
+				return err
+			}
+		}
+
+		// if hookName exists in the hook directory, run it
+		hookScript := filepath.Join(hooksDir, hookName)
+		_, err = os.Stat(hookScript)
+		if err == nil {
+			if stateMachine.commonFlags.Debug {
+				fmt.Printf("Running hook script: %s\n", hookScript)
+			}
+			if err := helper.RunScript(hookScript); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// handleLkBootloader handles the special "lk" bootloader case where some extra
+// files need to be added to the bootfs
+func (stateMachine *StateMachine) handleLkBootloader(volume *gadget.Volume) error {
+	if volume.Bootloader != "lk" {
+		return nil
+	}
+	// For the LK bootloader we need to copy boot.img and snapbootsel.bin to
+	// the gadget folder so they can be used as partition content. The first
+	// one comes from the kernel snap, while the second one is modified by
+	// the prepare_image step to set the right core and kernel for the kernel
+	// command line.
+	bootDir := filepath.Join(stateMachine.tempDirs.unpack,
+		"image", "boot", "lk")
+	gadgetDir := filepath.Join(stateMachine.tempDirs.unpack, "gadget")
+	if _, err := os.Stat(bootDir); err != nil {
+		return fmt.Errorf("got lk bootloader but directory %s does not exist", bootDir)
+	}
+	err := osMkdir(gadgetDir, 0755)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Failed to create gadget dir: %s", err.Error())
+	}
+	files, err := ioutilReadDir(bootDir)
+	if err != nil {
+		return fmt.Errorf("Error reading lk bootloader dir: %s", err.Error())
+	}
+	for _, lkFile := range files {
+		srcFile := filepath.Join(bootDir, lkFile.Name())
+		if err := osutilCopySpecialFile(srcFile, gadgetDir); err != nil {
+			return fmt.Errorf("Error copying lk bootloader dir: %s", err.Error())
+		}
+	}
+	return nil
+}
+
+// shouldSkipStructure returns whether a structure should be skipped during certain processing
+func shouldSkipStructure(structure gadget.VolumeStructure, isSeeded bool) bool {
+	if isSeeded &&
+		(structure.Role == gadget.SystemBoot ||
+			structure.Role == gadget.SystemData ||
+			structure.Role == gadget.SystemSave ||
+			structure.Label == gadget.SystemBoot) {
+		return true
+	}
+	return false
+}
+
+// copyStructureContent handles copying raw blobs or creating formatted filesystems
+func (stateMachine *StateMachine) copyStructureContent(structure gadget.VolumeStructure,
+	contentRoot, partImg string) error {
+	if structure.Filesystem == "" {
+		// copy the contents to the new location
+		var runningOffset quantity.Offset = 0
+		for _, content := range structure.Content {
+			if content.Offset != nil {
+				runningOffset = *content.Offset
+			}
+			// first zero it out. Structures without filesystem specified in the gadget
+			// yaml must have the size specified, so the bs= argument below is valid
+			ddArgs := []string{"if=/dev/zero", "of=" + partImg, "count=0",
```

Comment:
This is a minor thing: I'd like us to decide on the formatting style for how we concatenate strings. In earlier parts of the code we had no whitespaces, e.g. `"of="+partImg`, but now we do. I'm fine either way - I guess I always was more of a 'whitespace-yes' type of person.
In case we decide we want to have whitespaces, I'd recommend doing that cleanup with other lines of code in a separate PR.

Reply:
This is odd. I'm also a "whitespace-yes" type of person, but `go fmt` disagrees with us and usually removes the whitespace around `+` in string concatenations.. I'm curious why `go fmt` left this one... maybe because it's inside brackets? I'll look into it.

Reply:
Weird, `go fmt` is actually forcing them to be inconsistent like this...